src/server/pps/server/api_server.go

```diff
@@ -1833,6 +1833,16 @@ func (a *apiServer) hardStopPipeline(pachClient *client.APIClient, pipelineInfo
 	); err != nil && !isNotFoundErr(err) {
 		return errors.Wrapf(err, "could not recreate original output branch")
 	}
+	if pipelineInfo.EnableStats {
+		if err := pachClient.CreateBranch(
+			pipelineInfo.Pipeline.Name,
+			"stats",
+			"stats",
+			nil,
+		); err != nil && !isNotFoundErr(err) {
+			return errors.Wrapf(err, "could not recreate original stats branch")
+		}
+	}
 
 	// Now that new commits won't be created on the master branch, enumerate
 	// existing commits and close any open ones.
```

Comment:
Does this logic (closing any existing open commits) need to be applied to the stats branch as well?

Reply:
Yeah, I thought about adding some code to do that, but I was thinking we could lean on the recovery code for the time being since there are a few other places where we only finish the output commit to terminate a job. But, I guess that won't work in the reprocess case since the commit chain is broken. So, we wouldn't be blocked anywhere, but some stats commits could be left open after we break the commit chain. Making this change could still result with job state not being updated correctly though. I've discussed this a few times with @jdoliner and @nitinjainsj , but I really think we need to go ahead and make the appropriate changes such that we finish all relevant job state / commits transactionally rather than having this indirection through pfs. So, it basically plays out like: finish commit -> notice commit is finished -> finish job / stats commit. A better model that would guarantee consistent state would just be a single finish / stop job that handles it all transactionally. It is pretty easy to map an output commit to a job, but since jobs may not always exist for an output commit, that may make things a bit trickier.

Reply:
Sure—as an intermediate fix, we could change FinishCommit so that if `empty = true` it also finishes all subvenant commits (also setting `empty = true`), all transactionally?

Reply:
After some thought, I think it makes sense to keep this fix simple for the time being to keep the risk of new issues low. In a 2.0 release I will add a more complete fix to this problem, then backport it when we are confident the problem is solved.

Reply:
Sure, that sounds good. LGTM still LGTM